source/common/router/upstream_request.cc

```diff
@@ -286,6 +295,12 @@ void UpstreamRequest::decodeHeaders(Http::ResponseHeaderMapPtr&& headers, bool e
 
   maybeHandleDeferredReadDisable();
   ASSERT(headers.get());
+
+  // According to RFC7231 any 2xx response indicates that the connection is established.
```

Comment:
Is 2xx the only case we want to mark it InProgress? How about other codes with a body? I'm not sure whether we want to depend on the status code.

Reply:
Setting InProgress is mainly to be able to distinguish access logs in mid streams (using periodic access logs). Therefore, I think that setting InProgress in cases the stream will closed after response status code might confuse a user.

There would be a follow up PR to add period upstream access logs (similar to HCM periodic upstream access logs), then consider the following case:
The HTTP response headers had returned with 400 status code, the stream state is set to InProgress, and then an access log is recorded due to timer triggered. The output will include InProgress state, although the stream is terminating.

WDYT?

Reply:
After an offline discussion, we can to an agreement that in some cases a very large response body or a network delay may cause a time gap between the headers response and finish of the body encoding, regardless of the HTTP status code. Therefore, removed the status code check and keeping only "!end_stream"